lib/charms/prometheus_k8s/v0/prometheus_remote_write.py

```diff
@@ -1108,18 +924,51 @@ def apply_label_matchers(self, rules):
                     if label in rule["labels"]:
                         topology[label] = rule["labels"][label]
 
-                rule["expr"] = self._apply_label_matcher(rule["expr"], topology)
+                rule["expr"] = self.inject_label_matchers(rule["expr"], topology)
         return rules
 
-    def _apply_label_matcher(self, expression, topology):
+    def validate_alert_rules(self, rules: dict) -> Tuple[bool, str]:
+        """Will validate correctness alert rules, returning a boolean and any errors."""
+        if not self.path:
+            logger.debug("`cos-tool` unavailable. Not validating alert correctness.")
+            return True, ""
+
+        with tempfile.TemporaryDirectory() as tmpdir:
+            rule_path = Path(tmpdir + "/validate_rule.yaml")
+
+            # Smash "our" rules format into what upstream actually uses, which is more like:
+            #
+            # groups:
+            #   - name: foo
+            #     rules:
+            #       - alert: BlahBlah
+            #         expr: up
+            #       - alert: Haha
+            #         expr: up
+            transformed_rules = {"groups": []}  # type: ignore
+            for rule in rules["groups"]:
+                transformed = {"name": str(uuid.uuid4()), "rules": [rule]}
+                transformed_rules["groups"].append(transformed)
+
+            rule_path.write_text(yaml.dump(rules))
+
+            args = [str(self.path), "v", str(rule_path)]
+            # noinspection PyBroadException
+            try:
```

Comment:
Should the try-except be absorbed by `self._exec`?
```python
success, stdout = self._exec(args)
return success, stdout
```

Reply:
`stdout` and `stderr` (in `e.output`) are not equivalent and shouldn't be treated that way.

Reply:
That's exactly the same ;)

Success here is implicitly "didn't throw `CalledProcessError`". 

If the return code is 0, there's no output at all. If the return code is 1, an exception must be caught and stderr (from `e`) must be read.

It's not the same semantics as the return values from `transform`, so `_exec` can't do both without coupling it. Catching the exception is fine.